src/Illuminate/Database/Eloquent/Collection.php

```diff
@@ -83,7 +83,12 @@ public function contains($key, $operator = null, $value = null)
             return parent::contains(...func_get_args());
         }
 
-        $key = $key instanceof Model ? $key->getKey() : $key;
+        if ($key instanceof Model) {
+            return parent::contains(function ($model) use ($key) {
+                return $model->getKey() == $key->getKey() &&
+                       get_class($model) == get_class($key);
```

Comment:
Can we use `$model->is($key)` instead if possible? That has some additional checks such as connection as well!